src/stash/src/upgrade/v22_to_v23.rs

```diff
@@ -0,0 +1,457 @@
+// Copyright Materialize, Inc. and contributors. All rights reserved.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+use crate::upgrade::MigrationAction;
+use crate::{StashError, Transaction, TypedCollection};
+
+pub mod objects_v22 {
+    include!(concat!(env!("OUT_DIR"), "/objects_v22.rs"));
+}
+pub mod objects_v23 {
+    include!(concat!(env!("OUT_DIR"), "/objects_v23.rs"));
+}
+
+const MZ_SYSTEM_ROLE_ID: objects_v23::RoleId = objects_v23::RoleId {
+    value: Some(objects_v23::role_id::Value::System(1)),
+};
+const MZ_INTROSPECTION_ROLE_ID: objects_v23::RoleId = objects_v23::RoleId {
+    value: Some(objects_v23::role_id::Value::System(2)),
+};
+const PUBLIC_ROLE_ID: objects_v23::RoleId = objects_v23::RoleId {
+    value: Some(objects_v23::role_id::Value::Public(objects_v23::Empty {})),
+};
+const ACL_MODE_USAGE: objects_v23::AclMode = objects_v23::AclMode { bitflags: 1 << 8 };
+
+/// Give mz_introspection USAGE privileges on all clusters.
+pub async fn upgrade(tx: &'_ mut Transaction<'_>) -> Result<(), StashError> {
+    const DEFAULT_PRIVILEGES_COLLECTION: TypedCollection<
+        objects_v22::DefaultPrivilegesKey,
+        objects_v22::DefaultPrivilegesValue,
+    > = TypedCollection::new("default_privileges");
+    const CLUSTER_COLLECTION: TypedCollection<objects_v22::ClusterKey, objects_v22::ClusterValue> =
+        TypedCollection::new("compute_instance");
+
+    DEFAULT_PRIVILEGES_COLLECTION
+        .migrate_to(tx, |entries| {
+            let mut updates = vec![MigrationAction::Insert(
+                objects_v23::DefaultPrivilegesKey {
+                    role_id: Some(PUBLIC_ROLE_ID),
+                    database_id: None,
+                    schema_id: None,
+                    object_type: objects_v23::ObjectType::Cluster.into(),
+                    grantee: Some(MZ_INTROSPECTION_ROLE_ID),
+                },
+                objects_v23::DefaultPrivilegesValue {
+                    privileges: Some(ACL_MODE_USAGE),
+                },
+            )];
+
+            for (key, value) in entries {
+                let new_key = key.clone().into();
+                let new_value = value.clone().into();
+                updates.push(MigrationAction::Update(key.clone(), (new_key, new_value)));
+            }
+
+            updates
+        })
+        .await?;
+
+    CLUSTER_COLLECTION
```

Comment:
Should we be updating the `CLUSTER_COLLECTION` in the Stash initialization too?

Reply:
Yes, good catch. Just pushed an update.